### ISSUE TITLE
Default Keras optimizers using XLA causes error with TensorFlow-DirectML plugin

### ISSUE THREAD

**Issue by maggie1059** [keras-team/keras#17325]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): win64, WSL
- TensorFlow installed from (source or binary): tensorflow-directml-plugin (PyPI)
- TensorFlow version (use command below): tensorflow-cpu==2.11
- Exact command to reproduce: python keras/engine/training_test.py

**Describe the problem**.

Due to the refactoring of optimizers in Keras 2.11, default optimizers created for Keras models using model.compile() default to using jit_compile=True, and this is only turned off if no GPU devices are found. However, DirectML does not support jit compile / XLA, and so the program will throw an error because it detects a GPU, even though the GPU doesn't support XLA.

**Describe the current behavior**.

The model will throw an error when the optimizer tries to [apply gradients](https://github.com/keras-team/keras/blob/a4ed7cbe53706535cc930e45ba0969afa7f89d71/keras/optimizers/optimizer.py#L1229):
`Could not find compiler for platform DML: NOT_FOUND: could not find registered compiler for platform DML -- was support for that platform linked in?`

**Describe the expected behavior**.

If a non-XLA-capable GPU is detected, optimizers/models should fall back to non-XLA instead of throwing an error. Currently, a workaround is to use the experimental optimizers instead, which have the option to specify jit_compile=False. However, the defaults should ideally work with TensorFlow plugins without having to use this kind of workaround.

**Standalone code to reproduce the issue**.

`pip install tensorflow-directml-plugin`
`python keras/engine/training_test.py`

**Comment by SuryanarayanaY** [comment]
Hi @maggie1059 ,
I tried to replicate the issue in colab file.I did import tensorflow-cpu==2.11 but could not find the file `training_test.py `in the path.I could not able find it in 2.11V or nightly from source code.Could you help me with this file ? I tried testing on `!python training_v1.py` and `!python training.py` and throws no error.Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/08024a7d9677aba32c8768e64ea147bb/17325.ipynb).

I see this code [here](https://github.com/keras-team/keras/blob/a4ed7cbe53706535cc930e45ba0969afa7f89d71/keras/optimizers/optimizer.py#L57-60).

>         if not tf.config.list_physical_devices("GPU"):
>             # Optimizer only benefits from XLA when training on GPU. So if no
>             # GPU is found, we turn off XLA.
>             self.jit_compile = False

Even though default `jit_compile=True` at  `__init__` the above code makes `jit_compile=False` whenever GPU not found.

However with Tensorflow-DirectML plugin we need to check the behaviour but the blocker is the path `python keras/engine/training_test.py` .Whether we can find this path only when we use DirectML plugin ? Please confirm.

Thankyou!

**Comment by PatriceVignola** [comment]
Hi @SuryanarayanaY, I'm working with @maggie1059 on tensorflow-directml-plugin. I think the file she was referring to is this one: https://github.com/keras-team/keras/blob/master/keras/engine/training_test.py

The problem here is the assumption that a device with the `GPU` type is necessarily a `CUDA`, which isn't necessarily true anymore when pluggable devices are used. Our tensorflow-directml-plugin pluggable device uses the `GPU` string even though it has no `XLA` support, which causes the failure. I believe other pluggable devices also do the same thing (e.g. Apple's Metal pluggable device uses the `GPU` string as well).

Is there a way to detect that the device supports `XLA`? For example, maybe you could check that the `XLA_GPU` devices exist? TensorFlow seems to be exposing useful functions to check whether the binary is compiled with CUDA/ROCM/XLA: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/port_wrapper.cc

I believe a combination of those functions would be enough to keep the existing behavior, while also enabling pluggable devices used in conjunction with tensorflow-cpu to keep working. What do you think?

Thank you!

**Comment by SuryanarayanaY** [comment]
@gowthamkpr , Could you please look into the issue ? Thanks.

**Comment by sampathweb** [comment]
@PatriceVignola @maggie1059 - Thanks for reporting this issue.  We had similar issue for Mac M1 and TF 2.11 and has been fixed with this change to not `jit__compile` with XLA with this change - https://github.com/keras-team/keras/blob/master/keras/utils/tf_utils.py#L682
But it doesn't address your issue.  I have been looking at `tf.test.is_built_with_xla()` function, but waiting on confirmation and might be slightly delayed due to holidays.

**Comment by rjdyk** [comment]
Hi @sampathweb, do you have any updates on the status of this issue? The release of our DirectML PluggableDevice is blocked on this so we'd love to have some more clarity on this.

**Comment by rjdyk** [comment]
Hi @sampathweb, @fchollet -- could we get some eyes on this to check in the bug fix for the 2.12 release?

**Comment by rjdyk** [comment]
Hi @sampathweb. Do you have any updates on the status of this bug? Users of our DirectML PluggableDevice are running into issues due to this and we need this resolved for the 2.12 release of TensorFlow/Keras. 
I know the TensorFlow 2.12 branch cut happens this week. What timeline will Keras follow for their next release?

**Comment by qlzh727** [comment]
@rjdyk, it seems that the issue is probably addressed [by](https://github.com/keras-team/keras/issues/17325#issuecomment-1363377033), can you run your tests with latest keras nightly and confirm if that's the case? 

I don't think we will cherrypick the fix to 2.11.

**Comment by maggie1059** [comment]
Hi @qlzh727, unfortunately we are still running into the same error. If you were referring to [this fix](https://github.com/keras-team/keras/commit/1b32391798a952176b733660c940b1589c2fc8a4), it looks like it's only addressing the case for Mac M1 (`tf_utils.can_jit_compile()` still returns `True` with our plugin despite not having XLA support). I believe a fix for our case would require something more along the lines of what @PatriceVignola [suggested](https://github.com/keras-team/keras/issues/17325#issuecomment-1352320463). Would it be possible to look into getting a fix for this in before the 2.12 branch is cut?

**Comment by qlzh727** [comment]
Thanks for the confirmation. Since the release cut is approaching, I am not sure if we have enough room to fix this.

As we suggested on the email thread, is it possible to fallback to use `tf.keras.optimizers.legacy.xxx` on your side?

**Comment by chenmoneygithub** [comment]
Sorry I wonder why `tf_utils.can_jit_compile()` would return differently when DirectML is installed, if DirectML forces disabling the XLA, we should be able to detect that at `tf_utils.can_jit_compile()` ?

**Comment by fchollet** [comment]
Can we detect DirectML? If so we can just add that condition to `can_jit_compile()`.

**Comment by fchollet** [comment]
For the record there's a `tf.test.is_built_with_xla()` utility. However it returns `True` on M1 devices and is therefore useless. I do not know what it returns on DirectML. I suggest we keep doing fine-grained detected of non-supported runtimes since we do not seem to have access to a generalizable solution (beyond a try/except, which would be potentially expensive)

**Comment by maggie1059** [comment]
We'd prefer for users to be able to use the default optimizers rather than limiting them to using the legacy optimizers with our plugin. Unfortunately TensorFlow doesn't expose the "DML" device name to the python API, only the "GPU" type, which is causing these utils that look for XLA to return `True`, since they just look for the "GPU" string. Would it be possible to try doing something like this:

```
from tensorflow.python import pywrap_tfe
def can_jit_compile():
    """Returns True if TensorFlow XLA is available for the platform."""
    if len(pywrap_tfe.TF_ListPluggablePhysicalDevices()) > 0:
        return False
    return True
```
    
The `pywrap_tfe.TF_ListPluggablePhysicalDevices()` function returns a list of devices (such as `[b'/physical_device:GPU:0']`) if any pluggable devices are installed, and an empty list if not. I believe XLA does not currently work with pluggable devices, so this would exclude XLA from "GPU" pluggable devices. This should work for Mac as well, so it should be a more general solution than the current `can_jit_compile`.

**Comment by chenmoneygithub** [comment]
@maggie1059 Thx! Your proosal could work, would you mind creating an issue including your proposal in TF github repo?  To me it seems the current `tf_utils.can_jit_compile()` is not considering this pluggable case, so we should fix it there.

**Comment by fchollet** [comment]
The proposal sounds good, I will try that.

**Comment by maggie1059** [comment]
@chenmoneygithub Please correct me if I'm mistaken, but I believe the `tf_utils.can_jit_compile()` function only exists [in the keras repo](https://github.com/keras-team/keras/blob/8f1327566d708f0a0164ec9d0d03adc06c885ed8/keras/utils/tf_utils.py#L695), not the TF repo?

**Comment by qlzh727** [comment]
FYI, https://github.com/keras-team/keras/commit/8f1327566d708f0a0164ec9d0d03adc06c885ed8 should address this issue and be included in the latest keras-nightly. Can u run a test for that?

**Comment by chenmoneygithub** [comment]
@maggie1059 Sorry you are right. Could you try the testing again? Thx!

**Comment by maggie1059** [comment]
Yes, looks like everything is working as expected now! Thank you for your help!

**Comment by rjdyk** [comment]
Hi @fchollet, can you confirm that this change is going to ship in the next version of keras?
